test/jest/Bladeburner/Actions.test.ts

```diff
@@ -0,0 +1,358 @@
+      });
+
+      /** Relates to all issues mentioned in PR-1586 */
+      it.each(CITIES)("MILDLY increases chaos in all cities when chaos is HIGH: %s", (city: CityName) => {
+        guaranteeSuccess(), allCitiesHighChaos(), ({ chaos } = inst.cities[city]), complete(Incite);
+        expect(inst.cities[city].chaos).toBeGreaterThan(chaos * 1.05);
+      });
+    });
+
+    describe(BladeburnerGeneralActionName.Recruitment, () => {
+      const Recruitment = GeneralAction.createId(BladeburnerGeneralActionName.Recruitment);
+
+      it("awards charisma exp", () => {
+        guaranteeSuccess(), (before = Player.exp.charisma), complete(Recruitment);
+        expect(Player.exp.charisma).toBeGreaterThan(before);
+      });
+
+      it("hires team member", () => {
+        guaranteeSuccess(), complete(Recruitment);
+        expect(inst.teamSize).toBeGreaterThan(0);
+      });
+    });
+
+    describe.each([...actionId(contracts())])("$id.name", ({ id }) => {
+      it("all contracts award money", () => {
+        guaranteeSuccess(), (before = Player.money), complete(id);
+        expect(Player.money).toBeGreaterThan(before);
+      });
+    });
+
+    /** Stat EXP check for all actions */
+    /** Checking all of them to avoid regressions */
+    describe.each([...actionIdWithIndividualStat(nonGeneralActions())])("$id.name", ({ id, stat }) => {
+      it(`awards ${stat} exp`, () => {
+        guaranteeSuccess(), (before = Player.exp[stat]), complete(id);
+        expect(Player.exp[stat]).toBeGreaterThan(before);
+      });
+    });
+  });
+
+  describe("Upon failed completion", () => {
+    let before;
+
+    describe.each([SampleOperation, SampleBlackOp])("operations and black operations decrease rank", (id) => {
+      it(`${id.type}`, () => {
+        guaranteeFailure(), (before = inst.rank), complete(id);
+        expect(inst.rank).toBeLessThan(before);
+      });
+    });
+  });
+
+  it("have a minimum duration of 1 second", () => {
+    complete(SampleContract);
+    expect(inst.actionTimeToComplete).toBeGreaterThanOrEqual(1);
+  });
+
+  beforeAll(() => {
+    /* Initialise Formatters. Dependency of Bladeburner Logs/Console */
+    FormatsNeedToChange.emit();
+  });
+
+  beforeEach(() => {
+    setPlayer(new PlayerObject());
+
+    /** Need BN5 to receive Int EXP */
+    Player.sourceFiles.set(5, 3);
+
+    if (initBladeburner(Player)) {
+      inst = Player.bladeburner;
+      inst.clearConsole();
+    }
+  });
+
+  function initBladeburner(player: PlayerObject): player is PlayerObject & { bladeburner: Bladeburner } {
+    player.startBladeburner();
+    return true;
+  }
+
+  function basicStats() {
+    inst.rank = 1;
+    inst.changeRank(Player, 10);
+    Player.gainStrengthExp(LOW_STAT_EXP);
+    Player.gainDefenseExp(LOW_STAT_EXP);
+    Player.gainAgilityExp(LOW_STAT_EXP);
+    Player.gainDexterityExp(LOW_STAT_EXP);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
```

Comment:
Player setup where we want to keep numbers small in order to more precisely measure floating point operations (usually exp rewards).